src/v/jl/jls_writer/jls_writer.cc

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "jl/jls_writer/jls_writer.h"
+
+#include "base/likely.h"
+#include "base/vassert.h"
+#include "bytes/iobuf.h"
+#include "bytes/iostream.h"
+#include "jl/jls/jls.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_types.h"
+#include "serde/rw/rw.h"
+
+#include <seastar/core/future.hh>
+
+#include <__utility/move.h>
+
+#include <stdexcept>
+#include <stdint.h>
+#include <utility>
+
+namespace {
+void write_ntp(iobuf& buf, const model::ntp& ntp) {
+    serde::write(buf, ntp.ns);
+    serde::write(buf, ntp.tp.topic);
+    serde::write(buf, ntp.tp.partition);
+}
+} // namespace
+
+namespace jl::jls_writer {
+
+ss::future<> writer::write(model::ntp ntp, model::record_batch&& batch) {
+    vassert(
+      batch.header().type == model::record_batch_type::raft_data,
+      "Only raft data expected in the jls writer");
+
+    if (unlikely(_closed)) {
+        throw std::runtime_error("writer is closed");
+    }
+
+    if (unlikely(!_did_write_header)) {
+        co_await write_header();
+    }
+
+    if (ntp != _last_ntp || _chunk_buffer.empty()) {
+        if (unlikely(ntp < _last_ntp)) {
+            throw std::runtime_error(
+              "writer api invariant violation: ntps are out of order");
+        }
+
+        maybe_flush_batch_buffer();
```

Comment:
let's not do the buffering here
the buffering is already done in the uploader, the code should just serialize stuff into the `iobuf`
and it's totally OK to expect that the caller will serialize different ntps in correct order one by one